### PR TITLE
Add BERT Model To NeMo2.0

### PR DESCRIPTION
**Pull request by suiyoubi**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/bert/model/google_bert.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+        if self.type == 'model':
+            source = BertModel.from_pretrained(str(self), torch_dtype='auto')
+        elif self.type == 'pretraining':
+            source = BertForPreTraining.from_pretrained(str(self), torch_dtype='auto')
+        elif self.type == 'masked':
+            source = BertForMaskedLM.from_pretrained(str(self), torch_dtype='auto')
+        elif self.type == 'classification':
+            source = BertForNextSentencePrediction.from_pretrained(str(self), torch_dtype='auto')
+
+        print(
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/google_bert.py:92]: logging instead?

File: nemo/collections/llm/bert/model/google_bert.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+            f"Initializing Bert Model with pooler={self.config.add_pooler} "
+            f"lm_head={self.config.add_lm_head}  binary_head={self.config.bert_binary_head}"
+        )
+        target = self.init()
+        trainer = self.nemo_setup(target)
+
+        breakpoint()
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/google_bert.py:99]: remove breakpoint

File: nemo/collections/llm/bert/model/google_bert.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+        self.convert_state(source, target)
+        self.nemo_save(output_path, trainer)
+
+        print(f"Converted Bert model to Nemo, model saved to {output_path}")
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/google_bert.py:103]: same

File: nemo/collections/llm/bert/model/google_bert.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+    if padded_emb_size > emb_size:
+        zeros_to_add = torch.zeros(
+            padded_emb_size - emb_size,
+            embedding.size(1),
+            dtype=embedding.dtype,
+            device=embedding.device,
+        )
+        # Concatenate the two tensors along rows
+        padded_embedding = torch.cat((embedding, zeros_to_add), dim=0)
+        print(padded_embedding.size())
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/google_bert.py:278]: remove

File: nemo/collections/llm/bert/model/google_bert.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+@io.state_transform(
+    source_key=("bert.embeddings.word_embeddings.weight",),
+    target_key="embedding.word_embeddings.weight",
+)
+def _import_embedding(ctx: io.TransformCTX, embedding):
+    divisible = ctx.target.config.make_vocab_size_divisible_by
+    emb_size = embedding.size(0)
+    padded_emb_size = int(math.ceil(emb_size / divisible) * divisible)
+    print(padded_emb_size, divisible, emb_size)
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/google_bert.py:268]: remove

File: nemo/collections/llm/bert/model/google_bert.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+@io.state_transform(
+    source_key=("embeddings.word_embeddings.weight",),
+    target_key="embedding.word_embeddings.weight",
+)
+def _import_embedding_2(ctx: io.TransformCTX, embedding):
+    divisible = ctx.target.config.make_vocab_size_divisible_by
+    emb_size = embedding.size(0)
+    padded_emb_size = int(math.ceil(emb_size / divisible) * divisible)
+    print(padded_emb_size, divisible, emb_size)
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/google_bert.py:355]: remove

File: nemo/collections/llm/bert/model/google_bert.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+    if padded_emb_size > emb_size:
+        zeros_to_add = torch.zeros(
+            padded_emb_size - emb_size,
+            embedding.size(1),
+            dtype=embedding.dtype,
+            device=embedding.device,
+        )
+        # Concatenate the two tensors along rows
+        padded_embedding = torch.cat((embedding, zeros_to_add), dim=0)
+        print(padded_embedding.size())
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/google_bert.py:365]: remove

File: nemo/collections/llm/bert/model/base.py [hunk 8]
```diff
@@ -0,0 +1,647 @@
...
+            assert (
+                self.num_layers // p_size
+            ) % vp_size == 0, "Make sure the number of model chunks is the same across all pipeline stages."
+        from megatron.core import parallel_state
+
+        transformer_layer_spec = self.transformer_layer_spec
+        if not isinstance(transformer_layer_spec, ModuleSpec):
+            transformer_layer_spec = transformer_layer_spec(self)
+
+        breakpoint()
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/base.py:150]: remove

File: nemo/collections/llm/bert/model/base.py [hunk 9]
```diff
@@ -0,0 +1,647 @@
...
+        if self.bert_type == 'megatron':
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/base.py:151]: I recommend removing this if else, and directly get `num_tokentypes` from the config directly, or maybe only override it if `self.bert_binary_head` is set to False.  The reason is RoBERTA uses  `num_tokentypes=1`, and I assume we want to use the same base model for it

File: nemo/collections/llm/bert/model/google_bert.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+from torch import nn
+
+from nemo.collections.common.tokenizers import AutoTokenizer
+from nemo.collections.llm.bert.model.base import BertConfig, BertModel
+from nemo.collections.llm.utils import Config
+from nemo.lightning import OptimizerModule, io, teardown
+
+
+@dataclass
+class GoogleBertConfig(BertConfig):
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/google_bert.py:17]: Just curious about the naming, are there different Bert Base or Large than Google's?

File: nemo/collections/llm/bert/model/base.py [hunk 11]
```diff
@@ -0,0 +1,647 @@
...
+                size=(embeddings.shape[0], embeddings.shape[2]),
+                dtype=torch.float32,
+                device=torch.cuda.current_device(),
+            )
+            for i, (embedding, mask) in enumerate(zip(embeddings, masks)):
+                output[i, :] = torch.mean(embedding[1 : mask - 1], dim=0)
+            return output
+
+        # logits and loss
+        breakpoint()
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/base.py:282]: remove

**Review by JRD971000 — COMMENTED** [review]
Left a few comments.

File: nemo/collections/llm/bert/model/google_bert.py (hunk 10 above)
Comment (suiyoubi) [reply nemo/collections/llm/bert/model/google_bert.py:18]: I don't think so, but I also not sure what could be a better naming ? Any suggestion here ?

Comment (JRD971000) [reply nemo/collections/llm/bert/model/google_bert.py:18]: Just BertBase and BertLarge?

Comment (suiyoubi) [reply nemo/collections/llm/bert/model/google_bert.py:18]: sure I'll rename those

**Review by JRD971000 — APPROVED** [review]
LGTM! Thanks a lot @suiyoubi
